rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+        method: The method to apply the spec checking to.
+        tag: The tag of the spec to check. Either "input" or "output". This is used
+        internally to defined an internal cache storage attribute based on the tag.
+
+    Returns:
+        True if the spec should be validated, False otherwise.
+    """
+    cache_store = getattr(cls_instance, f"__checked_{tag}_specs_cache__", None)
+    return cache_store is None or method.__name__ not in cache_store
+
+
+def _validate(
+    *,
+    cls_instance: object,
+    method: Callable,
+    data: Mapping[str, Any],
+    spec: Spec,
+    filter: bool = False,
+    tag: str = "input",
+) -> NestedDict:
+    """Validate the data against the spec.
+
+    Args:
+        cls_instance: The class instance that the method belongs to.
+        method: The method to apply the spec checking to.
+        data: The data to validate.
+        spec: The spec to validate against.
+        filter: If True, the data will be filtered to only include the keys that are
+            specified in the spec.
+        tag: The tag of the spec to check. Either "input" or "output". This is used
+            internally to defined an internal cache storage attribute based on the tag.
+
+    Returns:
+        The data, filtered if filter is True.
+    """
+    is_mapping = isinstance(spec, SpecDict)
+    cache_miss = _should_validate(cls_instance, method, tag=tag)
+
+    if is_mapping:
+        if not isinstance(data, Mapping):
+            raise ValueError(f"{tag} must be a Mapping, got {type(data).__name__}")
+        if cache_miss or filter:
+            data = NestedDict(data)
+
+    if cache_miss:
+        try:
+            spec.validate(data)
+        except ValueError as e:
+            raise ValueError(
+                f"{tag} spec validation failed on "
+                f"{cls_instance.__class__.__name__}.{method.__name__}, {e}."
+            )
+
+    return data
+
+
+@DeveloperAPI(stability="alpha")
+def check_input_specs(
+    input_spec: str,
+    *,
+    filter: bool = False,
+    cache: bool = False,
+):
+    """A general-purpose spec checker decorator for Neural Network base classes.
+
+    This is a stateful decorator
+    (https://realpython.com/primer-on-python-decorators/#stateful-decorators) to
+    enforce input specs for any instance method that has an argument named
+    `input_data` in its args.
+
+    It also allows you to filter the input data dictionary to only include those keys
+    that are specified in the model specs. It also allows you to cache the validation
+    to make sure the spec is only validated once in the entire lifetime of the instance.
+
+    Examples (See more exmaples in ../tests/test_specs_dict.py):
+
+        >>> class MyModel(nn.Module):
+        ...     @property
+        ...     def input_spec(self):
+        ...         return {"obs": TensorSpec("b, d", d=64)}
+        ...
+        ...     @check_input_specs("input_spec")
```

Comment:
can we use both `@check_input_specs` and `@check_output_specs` on the same function?

Reply:
Yes.